### PR TITLE
start, commit, roll back and cancel edit for mesh layer

### PR DESCRIPTION
**Pull request by vcloarec**
core and UI implementation of:
- start editing
- commit edits
- roll back edits
- cancel edits
- save edits

### CODE REVIEW

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -11154,7 +11156,22 @@ void QgisApp::toggleEditing()
...
 bool QgisApp::toggleEditing( QgsMapLayer *layer, bool allowCancel )
 {
-  QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( layer );
+  switch ( layer->type() )
+  {
+    case QgsMapLayerType::VectorLayer:
+      return toggleEditingVectorLayer( qobject_cast<QgsVectorLayer *>( layer ), allowCancel );
+      break;
+    case QgsMapLayerType::MeshLayer:
+      return toggleEditingMeshLayer( qobject_cast<QgsMeshLayer *>( layer ) );
```
Comment (PeterPetrik) [inline src/app/qgisapp.cpp:11165]: we do not allow cancel?

Comment (PeterPetrik) [reply src/app/qgisapp.cpp:11165]: even if we do not support it ATM, I would add it to the toggleEditingMeshLayer as arg and just not use it ATM?

File: src/core/mesh/qgsmeshdataprovider.h [hunk 2]
```diff
@@ -415,6 +427,24 @@ class CORE_EXPORT QgsMeshDataProvider: public QgsDataProvider, public QgsMeshDat
...
 
+
+    /**
+     * Returns the mesh driver metadata of the provider
+     *
+     * \return the mesh driver metadata of the provider
+     *
+     * \since QGIS 3.22
+     */
+    virtual QgsMeshDriverMetadata driverMetaData()  const;
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshdataprovider.h]: driverMetaData -> driverMetadata

File: src/core/mesh/qgsmeshdataprovider.h [hunk 3]
```diff
@@ -415,6 +427,24 @@ class CORE_EXPORT QgsMeshDataProvider: public QgsDataProvider, public QgsMeshDat
...
+
+
+    /**
+     * Closes the dataprovider and free every ressources used
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshdataprovider.h]: data provider ... resources

**Review by PeterPetrik — COMMENTED** [review]
looks good feel free to do official PR. not sure if you added some automatic tests?

File: src/app/qgisapp.cpp (hunk 1 above)
Comment (vcloarec) [reply src/app/qgisapp.cpp:11165]: we always allow cancel, that why we do not need this flag.

Comment (vcloarec) [reply src/app/qgisapp.cpp:11165]: but indeed we have to not allow cancel and add it

**Pull request closed without merging**